src/Algebra/Graph/AdjacencyMap/Algorithm.hs

```diff
@@ -222,9 +236,22 @@ reachable x = concat . bfs [x]
 -- 'isJust' . topSort                      == 'isAcyclic'
 -- @
 topSort :: Ord a => AdjacencyMap a -> Maybe [a]
-topSort m = if isTopSortOf result m then Just result else Nothing
-  where
-    result = Typed.topSort (Typed.fromAdjacencyMap m)
+topSort g = check $ execState (explore $ vertexList g) (Set.empty,[]) where
+   -- todo add early exit if cycle detected. also add ability to detect
```

Comment:
It would be nice if we could have `topSort :: AdjacencyMap a -> Either [a] [a]` where the `Left` case returns a cycle :)

Reply:
Done! (well, first draft)